### PR TITLE
Add robot account authn & authz implementation

### PR DESCRIPTION
**Pull request by wy65701436**
This commit is to add the jwt token service, and do the authn & authz for robot account.

Signed-off-by: wang yan <wangyan@vmware.com>

### CODE REVIEW

File: src/common/models/robot.go [hunk 1]
```diff
@@ -45,10 +46,10 @@ type RobotQuery struct {
...
 // RobotReq ...
 type RobotReq struct {
-	Name        string             `json:"name"`
-	Description string             `json:"description"`
-	Disabled    bool               `json:"disabled"`
-	Access      []*ResourceActions `json:"access"`
+	Name        string         `json:"name"`
+	Description string         `json:"description"`
+	Disabled    bool           `json:"disabled"`
+	Policy      []*rbac.Policy `json:"access"`
```
Comment (reasonerjt) [inline src/common/models/robot.go]: We call it `policy` in the struct but call it `access` in the request body, any reason for this inconsistency?

File: src/common/rbac/project/util.go [hunk 2]
```diff
@@ -30,6 +36,20 @@ var (
 	}
 )
 
+func policiesForPublicProjectRobot(namespace rbac.Namespace) []*rbac.Policy {
```
Comment (reasonerjt) [inline src/common/rbac/project/util.go]: `policiesForPublicProjectRobot`,`policiesForPublicProject`,`policiesForSystemAdmin `

The code looks extremely repetitive.

File: src/common/token/htoken.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+	"crypto/ecdsa"
+	"crypto/rsa"
+	"errors"
+	"fmt"
+	"github.com/dgrijalva/jwt-go"
+	"github.com/goharbor/harbor/src/common/utils/log"
+	"time"
+)
+
+// HToken ...
```
Comment (reasonerjt) [inline src/common/token/htoken.go]: what is an HToken?  Please add more comment.

File: src/core/filter/security.go [hunk 4]
```diff
@@ -147,6 +153,50 @@ func (s *secretReqCtxModifier) Modify(ctx *beegoctx.Context) bool {
...
 }
 
+type robotAuthReqCtxModifier struct{}
+
+func (r *robotAuthReqCtxModifier) Modify(ctx *beegoctx.Context) bool {
+	robotName, robotTk, ok := ctx.Request.BasicAuth()
+	if !ok {
+		return false
+	}
+	log.Debug("got robot information via token auth")
```
Comment (reasonerjt) [inline src/core/filter/security.go]: This is misleading, this information will be printed even when the request contains basic auth info for a regular user.

File: src/core/filter/security.go [hunk 5]
```diff
@@ -147,6 +153,50 @@ func (s *secretReqCtxModifier) Modify(ctx *beegoctx.Context) bool {
...
+	}
+	log.Infof(fmt.Sprintf("got robot token header, %v", htk.Header))
+	// Do authn for robot account, as Harbor only stores the token ID, just validate the ID and disable.
+	robot, err := dao.GetRobotByID(htk.Claims.(*token.RobotClaims).TokenID)
+	if err != nil {
+		log.Errorf("failed to get robot %s: %v", robotName, err)
+		return false
+	}
+	if robot == nil {
+		log.Error("the token is not valid.")
```
Comment (reasonerjt) [inline src/core/filter/security.go]: It should be the ID of robot does not exist.

File: src/core/filter/security.go [hunk 6]
```diff
@@ -147,6 +153,50 @@ func (s *secretReqCtxModifier) Modify(ctx *beegoctx.Context) bool {
...
+	robotName, robotTk, ok := ctx.Request.BasicAuth()
+	if !ok {
+		return false
+	}
+	log.Debug("got robot information via token auth")
+	if !strings.HasPrefix(robotName, common.RobotPrefix) {
+		return false
+	}
+	rClaims := &token.RobotClaims{}
+	htk := &token.HToken{}
```
Comment (reasonerjt) [inline src/core/filter/security.go]: This line is not needed

File: src/common/token/claims.go [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+import (
+	"github.com/dgrijalva/jwt-go"
+	"github.com/goharbor/harbor/src/common/rbac"
+	"github.com/pkg/errors"
+)
+
+// RobotClaims implements the interface of jwt.Claims
+type RobotClaims struct {
+	jwt.StandardClaims
+	TokenID   int64          `json:"id"`
```
Comment (reasonerjt) [inline src/common/token/claims.go]: I think we need to revisit the structure of the token to make it more extensible, but this should not block this PR.

File: src/common/rbac/project/util.go (hunk 2 above)
Comment (reasonerjt) [inline src/common/rbac/project/util.go]: not needed.

File: src/common/rbac/project/robot.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+	// Index whether the robot is authenticated
+	IsAuthenticated() bool
+	// GetUsername returns the name of robot
+	GetUsername() string
+	// GetPolicy get the rbac policies from security context
+	GetPolicies() []*rbac.Policy
+}
+
+// robot implement the rbac.User interface for project robot account
+type robot struct {
```
Comment (reasonerjt) [inline src/common/rbac/project/robot.go]: This struct can be moved to the package where robot security context sits, and the robotContext is not needed.  We can reuse the robot security context in the struct.

File: src/common/token/htoken.go [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+type HToken struct {
+	jwt.Token
+}
+
+// NewWithClaims ...
+func NewWithClaims(claims *RobotClaims) *HToken {
```
Comment (reasonerjt) [inline src/common/token/htoken.go]: I'd prefer the claims be transparent to the api package so the api code just call NewRobotToken(id, projectID...)

File: src/common/token/htoken.go [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+			Issuer:    DefaultOptions.Issuer,
+		},
+	}
+	return &HToken{
+		Token: *jwt.NewWithClaims(DefaultOptions.SignMethod, rClaims),
+	}
+}
+
+// SignedString get the SignedString.
+func (htk *HToken) SignedString() (string, error) {
```
Comment (reasonerjt) [inline src/common/token/htoken.go]: This is essentially getting the raw token string, so maybe calling it `Raw`?

File: src/common/token/htoken.go [hunk 11]
```diff
@@ -0,0 +1,78 @@
+package token
```
Comment (reasonerjt) [inline src/common/token/htoken.go]: Seems there should has a way to check the whether the token is expired?

File: src/core/filter/security.go (hunk 4 above)
Comment (reasonerjt) [inline src/core/filter/security.go]: misleading log, remove it.

File: src/core/filter/security.go (hunk 6 above)
Comment (reasonerjt) [inline src/core/filter/security.go]: Should we check the expiration in this func?

File: src/core/filter/security.go (hunk 5 above)
Comment (reasonerjt) [inline src/core/filter/security.go]: Misleading log?

File: src/core/filter/security.go [hunk 12]
```diff
@@ -147,6 +153,50 @@ func (s *secretReqCtxModifier) Modify(ctx *beegoctx.Context) bool {
...
+	htk, err := token.ParseWithClaims(robotTk, rClaims)
+	if err != nil {
+		log.Errorf("failed to decrypt robot token, %v", err)
+		return false
+	}
+	log.Infof(fmt.Sprintf("got robot token header, %v", htk.Header))
```
Comment (reasonerjt) [inline src/core/filter/security.go]: Why print this log message?

**Comment by heww** [comment]
lgtm

**Pull request merged**